src/unlocked_ssl.cc

```diff
@@ -0,0 +1,254 @@
+
+    _ssl_module = PyImport_ImportModule("_ssl");
+    if(!_ssl_module) goto cleanup;
+
+    SSLSocketType = PyObject_GetAttrString(ssl_module, "SSLSocket");
+    if(!SSLSocketType) goto cleanup;
+
+    SSLWantReadError = PyObject_GetAttrString(_ssl_module, "SSLWantReadError");
+    if(!SSLWantReadError) goto cleanup;
+
+#if defined(_WIN32) || defined(__CYGWIN__)
+    // TODO: more DLL names?
+
+    HMODULE windows_openssl_handle = GetModuleHandle(TEXT("libssl-1_1.dll"));
+    if(!windows_openssl_handle) goto cleanup;
+
+    *(void**)&SSL_read_ex = GetProcAddress(windows_openssl_handle, "SSL_read_ex");
+    *(void**)&SSL_get_error = GetProcAddress(windows_openssl_handle, "SSL_get_error");
+    *(void**)&SSL_get_shutdown = GetProcAddress(windows_openssl_handle, "SSL_get_shutdown");
+#else
+    // Find library at "import ssl; print(ssl._ssl.__file__)"
+
+    _ssl_module_path = PyObject_GetAttrString(_ssl_module, "__file__");
+    if(!_ssl_module_path) goto error;
+
+    openssl_handle = dlopen(PyUnicode_AsUTF8(_ssl_module_path), RTLD_LAZY | RTLD_NOLOAD);
+    if(!openssl_handle) goto error;
+
+    *(void**)&SSL_read_ex = dlsym(openssl_handle, "SSL_read_ex");
+    *(void**)&SSL_get_error = dlsym(openssl_handle, "SSL_get_error");
+    *(void**)&SSL_get_shutdown = dlsym(openssl_handle, "SSL_get_shutdown");
+
+    error:
+    if (!openssl_linked() && openssl_handle) dlclose(openssl_handle);
+    Py_XDECREF(_ssl_module_path);
+#endif
+
+    cleanup:
+    Py_XDECREF(ssl_module);
+    Py_XDECREF(_ssl_module);
+    if (!openssl_linked()) {
+        Py_XDECREF(SSLWantReadError);
+        Py_XDECREF(SSLSocketType);
+    }
+}
+
+bool openssl_linked() {
+    return SSL_read_ex &&
+        SSL_get_error &&
+        SSL_get_shutdown &&
+        SSLWantReadError &&
+        SSLSocketType;
+}
+
+static PyObject* unlocked_ssl_recv_into_impl(PySSLSocket *self, Py_ssize_t len, Py_buffer *buffer) {
+    char *mem;
+    size_t count = 0;
+    size_t readbytes = 0;
+    int retval;
+    int sockstate;
+    _PySSLError err;
+    PyObject *sock = GET_SOCKET(self);
+
+    mem = (char *)buffer->buf;
+    if (len <= 0 || len > buffer->len) {
+        len = (int) buffer->len;
+        if (buffer->len != len) {
+            PyErr_SetString(PyExc_OverflowError,
+                            "maximum length can't fit in a C 'int'");
+            goto error;
+        }
+        if (len == 0) {
+            count = 0;
+            goto done;
+        }
+    }
+
+    if (sock != NULL) {
+        if (((PyObject*)sock) == Py_None) {
+            PyErr_SetString(PyExc_ValueError, "Underlying socket connection gone");
+            return NULL;
+        }
+        Py_INCREF(sock);
+    }
+
+    do {
+        Py_BEGIN_ALLOW_THREADS;
+        do {
+            retval = SSL_read_ex(self->ssl, mem + count, len, &readbytes);
+            if (retval <= 0) {
+                break;
+            }
+            count += readbytes;
+            len -= readbytes;
+        } while (len > 0);
```

Comment:
I suggested in another branch to add `&& readbytes >= 16384` because in some earlier testing I did it would mean that it was the last block available if there were less. Any opinions on that? I don't know if there are systems that could use smaller max size blocks.

Reply:
@mnightingale I think this got lost somewhere because I did program. It somewhere, it seems like a good idea since it's the openssl default.

Reply:
I did investigate it, but it doesn't fit the logic of the loops.

I believe there are a few TCP congestion protocols going on that result in the first segment being <16KB before quickly ramping up.

- SSL_read_ex in our case is always non-blocking so I would presume it's a very quick call if it has nothing more available
  - returns 1 on success - it read at least one segment
  - returns 0 on failure - may be retryable, call SSL_get_error to find out
- Briefly SSL_ERROR_WANT_READ means we tried to read but it failed because at least 1 segment wasn't yet available

In our python code we before we call unlocked_ssl_recv_into_impl we would poll the socket and know there is data on the underlying socket but there might not be enough to return anything yet.

To illustrate the path and why I believe there isn't anything to gain:

```py
do {
  Py_BEGIN_ALLOW_THREADS;
  do {
      retval = SSL_read_ex(self->ssl, mem + count, len, &readbytes);
      if (retval <= 0) {
          break;
      }
      count += readbytes;
      len -= readbytes;
  } while (len > 0);
  err = _PySSL_errno(retval == 0, self->ssl, retval); // sets err.ssl = SSL_get_error(...)
  Py_END_ALLOW_THREADS;
  self->err = err;

  if (count > 0) {
      break;
  }

  // A load of unused sockstate logic that we mostly don't use except detecting shutdown
  // I kept it so it matches the cpython code/PR
} while (err.ssl == SSL_ERROR_WANT_READ ||
         err.ssl == SSL_ERROR_WANT_WRITE);
```

I'm going to ignore the len calculation because just assume we aren't going to full our buffer so len > 0 is always true.

| iteration | retval | readbytes | count |
| --------- | ------ | --------- | ----- |
| 1         | 1      | 4096      | 4096  |
| 2         | 1      | 16384     | 20480 |
| 3         | 1      | 16384     | 36864 |
| 4         | 0      | 0         | 36864 |

retval <= 0 so break out of the inner loop
count > 0 so break out of the outer loop and return it to python

The only way you'd get past the count > 0 break would be if the first call failed, in which case the error is checked to determine if we should retry or return to python.
It's likely we'll get an SSL_ERROR_WANT_READ since we've already polled and know there is some data on the socket, therefore we'll retry until the first segment is available, and if the next loop after a success fails count will be > 0 and we'll return to python as above.

Basically, to have even gotten to this stage we know there is or will be very shortly some data.
The len > 0 condition is really just making sure we limit our reads to the size of the buffer, as soon as we receive something and there is nothing left to read immediately available we exit.

---

I get what you're saying in that if we've had a 16 KB in the past we could stop reading once we receive a smaller segment, however in my opinion the simplicity of the loop means once the data available is exhausted its going to return anyway.

For the above reasons I don't think there is an issue with reading a large amount of data - I think we've settled on a 1 MB buffer, we are releasing the GIL so other threads can do their work anyway and I think their is a lot of potential with splitting downloading across multiple threads (from my testing very few pools of connections are necessary and it'll depend on total connections - there is no point sharing 8 connections across 4 threads) because multiple downloaders can be reading their data simultaneously.

Quite possibly the maximum read per cycle and hence buffer per connection should be configurable in SAB but that only really helps very slow connections or very low RAM systems - as far as I can tell a whole article is still going to be buffered into memory which the 1MB default helps with, if it starts with a smaller buffer it's going to need to keep growing which is an expensive process.